### PR TITLE
Private registry cleanup job

### PR DESCRIPTION
**Pull request by thedadams**
Issues:
https://github.com/rancher/rancher/issues/36981
https://github.com/rancher/rancher/issues/37079
https://github.com/rancher/rancher/issues/37063

## Problem
Add feature flag for custom RKE1 node cleanup

It is common for users to cleanup their own custom nodes when deleting
them from Rancher. This has caused issue with jobs and/or pods being
left around or nodes not being deleted properly.

## Solution
This change adds a feature flag to turn the cleanup off globally if a
user doesn't need it.

## Problem
Refactor node-cleanup job implementation

It was determined that some of the code around creating and deleting the
node-cleanup job was producing unexpected results like non-nil errors
with empty error messages. In addition, the cleanup job is being
leftover in some case.

## Solution
This code makes the error messages more descriptive and adds retries to
try to ensure that the cleanup job gets deleted.

## Problem
Add private registry information to node cleanup job

The node cleanup job ignored cluster private registry information. This caused
issues when trying to remove nodes from a custom RKE1 cluster when private
registries were enforced by something like OPA Gatekeeper.

## Solution
Now, just as with cluster agent deployment, private registry information is
added whenever needed.

## Problem
If a pod is stuck because it has been scheduled on a node that node
longer exists, then deleting the job may not be enough to delete this
pod. This would cause the job deletion to hang forever.

## Solution
This change deletes the pod with gradePeriodSeconds of 0 to force its
deletion.

### CODE REVIEW

File: pkg/controllers/management/node/cleanup.go [hunk 1]
```diff
@@ -185,64 +198,77 @@ func (m *Lifecycle) waitUntilJobCompletes(userContext *config.UserContext, job *
...
 		}
+
+		return nil, m.waitUntilJobDeletes(userContext, &existingJob.Items[0])
 	}
 
 	meta := metav1.ObjectMeta{
-		GenerateName: "cattle-node-cleanup-",
-		Namespace:    "default",
+		GenerateName: "cattle-cleanup-node-",
+		Namespace:    namespace.System,
```
Comment (luthermonson) [inline pkg/controllers/management/node/cleanup.go:265]: i think we should leave this in default for "reasons". there are more cleanup jobs in default that would need to be changed if you stick with this choice.

**Review by luthermonson — DISMISSED** [review]

**Review by snasovich — APPROVED** [review]

**Review by paynejacob — DISMISSED** [review]

**Review by jakefhyde — APPROVED** [review]

**Review by luthermonson — APPROVED** [review]

**Comment by martyav** [comment]
@thedadams I'm a technical writer on the Rancher team -- a user requested documentation for `rke1-custom-node-cleanup` as it was accidentally left out of https://ranchermanager.docs.rancher.com/getting-started/installation-and-upgrade/installation-references/feature-flags#docusaurus_skipToContent_fallback. Presumably the default value is `true` as it's active on a fresh Rancher install of v2.7.2. Is it considered general availability or experimental?

**Comment by snasovich** [comment]
@martyav , you're correct that this flag defaults to `true` and it was added as part of the linked issue https://github.com/rancher/rancher/issues/37065 that has the rationale on why it got added. Right around the same time we've made multiple fixes to improve the stability of node cleanup jobs and this setting was viewed as "plan B" in case the fixes don't work on some configurations (so far the fixes has been good).

So, the setting does exist and I would not consider it "experimental" per se - just one that we should only recommend changing if the user understands the consequences of disabling it. The node cleanup job was introduced to prevent situations where removed nodes may attempt to rejoin the cluster after being booted up again.

Hope this helps.

cc: @thedadams

**Comment by martyav** [comment]
@snasovich Thank you, was able to update and do a general refresh on the page

**Pull request merged**